### PR TITLE
Python Frontend Documentation

### PR DESCRIPTION
**Pull request by alexnick83**
The purpose of the PR is to create analytical documentation on the [Python Language Reference](https://docs.python.org/3/reference/) features supported by the Python-Frontend.
Emphasis will be given on selected sections from the following chapters:

- [x]  2: Lexical Analysis
- [x]  6: Expressions
- [x]  7: Simple Statements
- [x]  8: Compound Statements

### CODE REVIEW

**Comment by tbennun** [comment]
@alexnick83 might be better to use tables for some of the expressions in the markdown file (when it's either "supported" or "unsupported", it would be much easier to read).

**Comment by alexnick83** [comment]
> @alexnick83 might be better to use tables for some of the expressions in the markdown file (when it's either "supported" or "unsupported", it would be much easier to read).

I agree, but I am thinking of addressing this in a second PR.

**Review by tbennun — CHANGES_REQUESTED** [review]
Looks very good, just a few comments and corrections.

File: Jenkinsfile [hunk 1]
```diff
@@ -73,6 +73,7 @@ tests/mpi_test.sh'''
             source /opt/intelFPGA_pro/19.1/hld/init_opencl.sh
             export PATH=~/.local/bin:$PATH
 export DACE_debugprint=1
+export DACE_compiler_default_data_types=C
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply Jenkinsfile:76]: Hmm, I don't want to touch the tests, especially since the issue manifests only with Intel's FPGA toolchain. The problem is quite simple. The default value for the flag is `Python`, i.e. integral data are interpreted to be int64. This leads to generating code that calls an Intel FPGA API method with `long long` arguments. However, this method only supports `int` parameters. Perhaps @definelicht and @TizianoDeMatteis have better insight.

File: dace/frontend/python/newast.py [hunk 2]
```diff
@@ -1970,46 +1971,116 @@ def visit_For(self, node: ast.For):
...
+            self.break_states.append([])
             laststate, first_loop_state, last_loop_state = self._recursive_visit(
                 node.body, 'for', node.lineno, extra_symbols=extra_syms)
             end_loop_state = self.last_state
 
             # Add loop to SDFG
             loop_cond = '>' if ((
                 pystr_to_symbolic(ranges[0][2]) < 0) == True) else '<'
-            self.sdfg.add_loop(
+            _, loop_guard, loop_end = self.sdfg.add_loop(
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/newast.py:1984]: The line change (extracting the `loop_guard` and the `loop_end`) has been done in order to be able to properly support `break` and `continue` (see lines 1990-2005). I am basically generating other interstates to simulate `break` and `continue` when needed. As far as I understand, PR #161 has nothing to do with this.

File: dace/frontend/python/newast.py [hunk 3]
```diff
@@ -1970,46 +1971,116 @@ def visit_For(self, node: ast.For):
...
+            out_edges = self.sdfg.out_edges(next_state)
+            for e in out_edges:
+                self.sdfg.remove_edge(e)
+            self.sdfg.add_edge(next_state, loop_end, dace.InterstateEdge())
+        self.loop_idx -= 1
+
+    def visit_Break(self, node: ast.Break):
+        if self.loop_idx < 0:
+            error_msg = "'break' is only supported inside for and while loops "
+            if self.nested:
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/newast.py:2045]: Yes, this works (just tried). It doesn't work (and hopefully returns an informative error msg) when `break` or `continue` is in a nested scope relative to their corresponding loop guard. The reason that such a thing is not supported has to do with how `break`/`continue` are currently implemented; with interstate edges. Such an approach doesn't work with nested scopes, since you cannot connect a state of the nested scope with a node/state of the outer scope. This is also why Maps are not supported (apart from the semantical issue).

File: dace/frontend/python/newast.py [hunk 4]
```diff
@@ -1970,46 +1971,116 @@ def visit_For(self, node: ast.For):
...
+                              " used in nested DaCe program calls to break out "
+                              " of loops of outer scopes)")
+            raise DaceSyntaxError(self, node, error_msg)
+        self.break_states[self.loop_idx].append(self.last_state)
+    
+    def visit_Continue(self, node: ast.Continue):
+        if self.loop_idx < 0:
+            error_msg = ("'continue' is only supported inside for and while "
+                         "loops ")
+            if self.nested:
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/newast.py:2056]: See above

File: dace/frontend/python/newast.py [hunk 5]
```diff
@@ -1970,46 +1971,116 @@ def visit_For(self, node: ast.For):
...
         # Add a guard state
         self._add_state('if_guard')
+        if (isinstance(node.test, ast.Compare) and
+                isinstance(node.test.left, ast.Subscript)):
+            cond = self.visit(node.test)
+            if cond in self.sdfg.arrays:
+                cond_dt = self.sdfg.arrays[cond]
+                if isinstance(cond_dt, data.Array):
+                    cond += '[0]'
+            cond_else = 'not ({})'.format(cond)
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/newast.py:2073]: The above will be executed only when the if-test is a comparison of the form `data < sth`, i.e. there is a data-dependency. This will generate a boolean scalar or a boolean array of size 1, which in turn will be used in the interstate edge condition. Therefore, 'not ({})'.format(cond)` should always work. I thought that `negate_expr` works only with symbolic expressions, that's why I didn't use it.

File: dace/frontend/python/newast.py [hunk 6]
```diff
@@ -2103,49 +2167,46 @@ def _add_assignment(self,
...
-                        '__i%d' % i: '%s:%s+1:%s' % (start, end, step)
-                        for i, (start, end, step) in enumerate(target_subset)
-                    }, {},
-                                             '__out = %s' % operand,
-                                             {'__out': memlet},
-                                             external_edges=True,
-                                             debuginfo=self.current_lineinfo)
+                if op_name:
+                    inp_memlet = {'__inp': Memlet.simple(
+                        op_name, '%s' % op_subset)}  # TODO: Why [0][0]???
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/newast.py]: I guess not, since nothing broke. I was only puzzled about why the subset was used in the first place. Anyway, I removed them.

File: dace/frontend/python/python_supported_features.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+However, there is currently limited support for strings and char/byte arrays.
+For example, it is not possible to instantiate an (u)int8 array with a string
+or byte literal.
+
+### 2.5 Operators
+The DaCe Python-Frontend supports all Python operators.
+The operators are only supported in the context of arithmetic/logical operations among
+scalar values and DaCe (NumPy-compatible) arrays. For example, it is not possible
+to concatenate 2 strings with the `+` operator.
+The `:=` operator (Named Expression) is parsed as an assignment statement.
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/python_supported_features.md:55]: Python 3.8 :trippy:

File: Jenkinsfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (TizianoDeMatteis) [reply Jenkinsfile:76]: Can we set it just for Intel tests?

File: dace/frontend/python/python_supported_features.md [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+- Slicings: Supported on DaCe/Numpy arrays
+- Calls: Supported for other DaCe programs, and a subset of methods from the
+DaCe and Numpy modules
+
+### 6.4 Await expression
+Unsupported  
+
+### 6.5 The power operator
+Supported. If the base is an integer and the exponent a signed integer, both
+operands are cast to float64 and the result is also of type float64.
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/python_supported_features.md:100]: Good point. But in any case, the current behavior of operators (especially power, div, floordiv, modulo) should be considered "placeholder". The reason is that there is a great mismatch in the behavior of those operators among Python, Numpy, and our C-based code generators. I plan to make a PR that introduce one or more flag that will allow the user to influence what the operators actually do.

File: Jenkinsfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexnick83) [reply Jenkinsfile:76]: This is what is currently being done (through the Jenkins configuration file). I think @tbennun meant to set this somehow specifically in the actual test (py) files.

File: dace/frontend/python/python_supported_features.md [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+## 7 Simple Statements
+
+### 7.1 Expression statements
+Partially supported, as described in the previous section. Python interactive
+mode is not supported.
+
+### 7.2 Assignment statements
+Assignment statements with single or multiple targets are supported, both with
+and without parentheses. Statements with starred targets are not supported.
+Targets may only be identifiers, and subscriptions/slices of Numpy arrays.
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/python_supported_features.md:149]: Probably, I will test and update accordingly.

File: dace/frontend/python/replacements.py [hunk 10]
```diff
@@ -239,6 +241,8 @@ def _simple_call(sdfg: SDFG,
                  func: str,
                  restype: dace.typeclass = None):
     """ Implements a simple call of the form `out = func(inp)`. """
+    if isinstance(inpname, (list, tuple)):  # TODO investigate this
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/replacements.py:244]: I somehow wrote some code, where `inpname` should be e.g., `"A"`, but the value passed was actually `["A"]`. It is a quick fix. I think it is benign in any case.

File: dace/frontend/python/python_supported_features.md [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+- global, nonlocal
+- class
+- try, except, finally
+- raise, yield, pass
+- import, from, as
+- assert, async, await, del
+
+#### 2.3.2 Reserved classes of identifiers
+Reserved class of Python identifiers are not supported. Furthermore, identifiers
+starting with double underscore (`__`) are reserved by the SDFG language.
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/python_supported_features.md:42]: I am referring also to the connector names. I have standardized the connector names generated from ops to be `__inX` and `__outY`. Maybe not the best solution, but it should be good advice to just avoid anything starting with `__`.

File: dace/frontend/python/python_supported_features.md [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+### 7.3 The assert statement
+Unsupported
+
+### 7.4 The pass statement
+Unsupported
+
+### 7.5 The del statement
+Unsupported
+
+### 7.6 The result statement
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/python_supported_features.md:166]: Yes, fixed.

File: Jenkinsfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (tbennun) [reply Jenkinsfile:76]: ok, we can keep it because it's an architecture requirement. IMO I think we should just add dtypes to `dace.symbol` definitions in the samples that break.

File: dace/frontend/python/python_supported_features.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/python_supported_features.md:149]: Brain freeze, this is what I mean above. This passes:
```python
@dace.program
def multiple_targets_parentheses(a: dace.float32[1]):
    (b, c) = (a, 2*a)
    return b, c


def test_multiple_targets_parentheses():
    a = np.zeros((1, ), dtype=np.float32)
    a[0] = np.pi
    b, c = multiple_targets_parentheses(a=a)
    assert(b[0] == np.float32(np.pi))
    assert(c[0] == np.float32(2) * np.float32(np.pi))
```

File: dace/frontend/python/python_supported_features.md [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+
+#### 2.3.2 Reserved classes of identifiers
+Reserved class of Python identifiers are not supported. Furthermore, identifiers
+starting with double underscore (`__`) are reserved by the SDFG language.
+
+### 2.4 Literals
+The DaCe Python-Frontend supports in general the same literals as Python.
+However, there is currently limited support for strings and char/byte arrays.
+For example, it is not possible to instantiate an (u)int8 array with a string
+or byte literal.
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/python_supported_features.md:48]: This will go into a new PR that will allow users to write stuff such as `a = dace.float32(0.8)`

File: dace/frontend/python/python_supported_features.md [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+- 2.1.3 Comments
+- 2.1.5 Explicit Line Joining
+- 2.1.6 Implicit Line Joining
+- 2.1.7 Blank Lines
+
+### 2.3 Identifiers and Keywords
+The DaCe Python-Frontend uses exactly the same rules for identifiers as normal Python.
+However, for a generated SDFG to compile to a target architecture, the identifiers used
+must be valid in the output language (e.g., C/C++). We currently provide no automatic
+transformation of identifiers used as variable names. Therefore, it is the responsibility
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/python_supported_features.md:21]: If the user uses a variable name that is valid in Python, but invalid in C++, the SDFG will not compile. We do not transform/convert identifier names automatically.

File: tests/python_frontend/line_structure_test.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+def test_comments():
+    N.set(128)
+    A = np.random.rand(N.get()).astype(np.float32)
+    B = np.zeros((N.get(),), dtype=np.float32)
+    comments(A, B)
+    assert np.allclose(A, B)
+
+
+@dace.program
```
[the remark this answers is not in the document]
Comment (tbennun) [reply tests/python_frontend/line_structure_test.py:24]: @alexnick83 Done.

File: dace/frontend/python/python_supported_features.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/frontend/python/python_supported_features.md:149]: @alexnick83 that's pretty cool. Is it already a test?

File: dace/frontend/python/newast.py [hunk 16]
```diff
@@ -1942,19 +1986,39 @@ def visit_For(self, node: ast.For):
...
-            }
+            # extra_syms = {
+            #     indices[0]:
+            #     symbolic.symbol(
+            #         indices[0],
+            #         dtypes.result_type_of(
+            #             infer_expr_type(ranges[0][0], self.sdfg.symbols),
+            #             infer_expr_type(ranges[0][1], self.sdfg.symbols),
+            #             infer_expr_type(ranges[0][2], self.sdfg.symbols)))
+            # }
```
Comment (tbennun) [inline dace/frontend/python/newast.py:2001]:
```suggestion
```

File: dace/subsets.py [hunk 17]
```diff
@@ -576,17 +576,29 @@ def compose(self, other):
         else:
             raise NotImplementedError
 
-    def squeeze(self):
+    def squeeze(self, ignore_indices: List[int] = []):
```
Comment (tbennun) [inline dace/subsets.py:579]: Mutable default parameters are not good. Needs `= None`  and `ignore_indices = ignore_indices or []` within the function.

File: dace/sdfg/propagation.py [hunk 18]
```diff
@@ -772,3 +772,85 @@ def propagate_memlet(dfg_state,
         new_memlet.num_accesses = 0
 
     return new_memlet
+
+
+def propagate_symbol(dfg_state,
```
Comment (tbennun) [inline dace/sdfg/propagation.py:777]: Is it possible to use the function from #394 instead? If so we should consider merging it, since the refactor there is nice for more uses (state propagation).

**Review by tbennun — CHANGES_REQUESTED** [review]

File: dace/frontend/python/python_supported_features.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/python_supported_features.md:149]: Yes, it is in the tests.

File: dace/frontend/python/memlet_parser.py [hunk 19]
```diff
@@ -136,6 +136,8 @@ def parse_memlet_subset(array: data.Data, node: Union[ast.Name, ast.Subscript],
     else:  # Use entire range
         subset = _ndslice_to_subset(ndslice)
 
+    if isinstance(subset, subsets.Indices):
+        subset = subsets.Range([(i, i, 1) for i in subset])
```
Comment (tbennun) [inline dace/frontend/python/memlet_parser.py:140]: we have `subsets.Range.from_indices` now. Separate PR though.

File: dace/frontend/python/newast.py [hunk 20]
```diff
@@ -2926,8 +2966,10 @@ def visit_Call(self, node: ast.Call):
             # Remove newly-defined symbols from arguments
             if mapping is not None:
                 symbols -= set(mapping.keys())
-            if len(symbols) > 0:
-                mapping = mapping or {}
+            # if len(symbols) > 0:
+            #     mapping = mapping or {}
+            # TODO: Why above the None fix was applied when there were symbols?
```
Comment (tbennun) [inline dace/frontend/python/newast.py:2971]: I don't remember why, but we definitely added a test for that, so if it passes it's fine with me.

**Review by tbennun — APPROVED** [review]
LGTM

File: dace/frontend/python/newast.py (hunk 20 above)
Comment (alexnick83) [reply dace/frontend/python/newast.py:2971]: It is probably somehow related to the `_add_access` changes (for a test I got an error because mapping was None).

File: dace/frontend/python/memlet_parser.py (hunk 19 above)
Comment (alexnick83) [reply dace/frontend/python/memlet_parser.py:140]: Cool, we can update it when we merge.

File: dace/sdfg/propagation.py (hunk 18 above)
Comment (alexnick83) [reply dace/sdfg/propagation.py]: Done
